RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,432 @@
+    yield y + z
+ }
+```
+
+Unfortunately, the naive desugaring of this can make it very easy to build a resulting chain of method calls which unintentionally ends up being very large:
+
+```fsharp
+ce.Combine(
+    ce.Combine(
+        ce.Apply(
+            ce.Apply(
+                ce.Apply(
+                    ce.Return(
+                        (fun x ->
+                            (fun y ->
+                                (fun z ->
+                                    x + y + z
+                                )
+                            )
+                        )),
+                    foo),
+                bar),
+            baz),
+        ce.Apply(
+            ce.Apply(
+                ce.Apply(
+                    ce.Return(
+                        (fun x ->
+                            (fun y ->
+                                (fun z ->
+                                    x + y
+                                )
+                            )
+                        )),
+                    foo),
+                bar),
+            baz)
+    ),
+    ce.Apply(
+        ce.Apply(
+            ce.Apply(
+                ce.Return(
+                    (fun x ->
+                        (fun y ->
+                            (fun z ->
+                                y + z
+                            )
+                        )
+                    )),
+                foo),
+            bar), 
+        baz)
+)
+```
+
+*N.B.* the size of the desugared expression grows with the product of the number of bindings introduced by the `let! ... and! ...` syntax and the number calls to `Combine` implied by the alternative cases.
+
+An attempt at a very smart desugaring which tries to cut down the resulting expression might, on the face of it, seem like a reasonable option. However, beyond the cost of analysing which values which are introduced by `let! ... and! ...` actually go on to be used, we must also consider the right-hand sides of the `let! ... and! ...` bindings and the pattern matching: Do we evaluated these once up front? Or recompute them in each alternative case at the leaf of the tree of calls to `Combine`? What if the expressions on the right-hand sides have side-effects, or the left-hand side utilises active patterns with side-effects? At that point we either make complex, unintuitive rules, or force the CE writer to be explicit. Continuing in the spirit of CEs generally being straightforward desugarings, we choose the latter make make the writer clearly state their desire.
+
+In order to keep things simple, then, we keep the canonical form introduced earlier, which forces precisely one `return` after a `let! ... and! ...`. However, we can still express alternative applicatives when using the `let! ... and! ...` syntax, we just need to use the same trick as with additional `let!`s and leave the scope of the canonical applicative syntax and therefore leave the additional constraints it places upon us:
```

Comment:
Upon reading this section, I think it needs to be shuffled around a bit:

1. Explain that you can use `yield`
1. Explain the restrictions (i.e., must follow a canonical form)
1. Explain why this is required

Additionally, I suggest renaming the section to either `Using yield` or `Using monoids`.